// Prelude/index.js
"use strict";
var Unit = {
    create: function (value) {
        return value;
    }
};
function LT() {

};
LT.value = new LT();
function GT() {

};
GT.value = new GT();
function EQ() {

};
EQ.value = new EQ();
function Semigroupoid($less$less$less) {
    this["<<<"] = $less$less$less;
};
function Category(__superclass_Prelude$dotSemigroupoid_0, id) {
    this["__superclass_Prelude.Semigroupoid_0"] = __superclass_Prelude$dotSemigroupoid_0;
    this.id = id;
};
function Show(show) {
    this.show = show;
};
function Functor($less$dollar$greater) {
    this["<$>"] = $less$dollar$greater;
};
function Apply($less$times$greater, __superclass_Prelude$dotFunctor_0) {
    this["<*>"] = $less$times$greater;
    this["__superclass_Prelude.Functor_0"] = __superclass_Prelude$dotFunctor_0;
};
function Applicative(__superclass_Prelude$dotApply_0, pure) {
    this["__superclass_Prelude.Apply_0"] = __superclass_Prelude$dotApply_0;
    this.pure = pure;
};
function Alternative($less$bar$greater, empty) {
    this["<|>"] = $less$bar$greater;
    this.empty = empty;
};
function Bind($greater$greater$eq, __superclass_Prelude$dotApply_0) {
    this[">>="] = $greater$greater$eq;
    this["__superclass_Prelude.Apply_0"] = __superclass_Prelude$dotApply_0;
};
function Monad(__superclass_Prelude$dotApplicative_0, __superclass_Prelude$dotBind_1) {
    this["__superclass_Prelude.Applicative_0"] = __superclass_Prelude$dotApplicative_0;
    this["__superclass_Prelude.Bind_1"] = __superclass_Prelude$dotBind_1;
};
function Num($percent, $times, $plus, $minus, $div, negate) {
    this["%"] = $percent;
    this["*"] = $times;
    this["+"] = $plus;
    this["-"] = $minus;
    this["/"] = $div;
    this.negate = negate;
};
function Eq($div$eq, $eq$eq) {
    this["/="] = $div$eq;
    this["=="] = $eq$eq;
};
function Ord(__superclass_Prelude$dotEq_0, compare) {
    this["__superclass_Prelude.Eq_0"] = __superclass_Prelude$dotEq_0;
    this.compare = compare;
};
function Bits($amp, $up, complement, shl, shr, zshr, $bar) {
    this["&"] = $amp;
    this["^"] = $up;
    this.complement = complement;
    this.shl = shl;
    this.shr = shr;
    this.zshr = zshr;
    this["|"] = $bar;
};
function BoolLike($amp$amp, not, $bar$bar) {
    this["&&"] = $amp$amp;
    this.not = not;
    this["||"] = $bar$bar;
};
function Semigroup($less$greater) {
    this["<>"] = $less$greater;
};
function cons(e) {  return function (l) {    return [e].concat(l);  };};
function showStringImpl(s) {  return JSON.stringify(s);};
function showNumberImpl(n) {  return n.toString();};
function showArrayImpl (f) {  return function (xs) {    var ss = [];    for (var i = 0, l = xs.length; i < l; i++) {      ss[i] = f(xs[i]);    }    return '[' + ss.join(',') + ']';  };};
function numAdd(n1) {  return function(n2) {    return n1 + n2;  };};
function numSub(n1) {  return function(n2) {    return n1 - n2;  };};
function numMul(n1) {  return function(n2) {    return n1 * n2;  };};
function numDiv(n1) {  return function(n2) {    return n1 / n2;  };};
function numMod(n1) {  return function(n2) {    return n1 % n2;  };};
function numNegate(n) {  return -n;};
function refEq(r1) {  return function(r2) {    return r1 === r2;  };};
function refIneq(r1) {  return function(r2) {    return r1 !== r2;  };};
function eqArrayImpl(f) {  return function(xs) {    return function(ys) {      if (xs.length !== ys.length) return false;      for (var i = 0; i < xs.length; i++) {        if (!f(xs[i])(ys[i])) return false;      }      return true;    };  };};
function unsafeCompareImpl(lt) {  return function (eq) {    return function (gt) {      return function (x) {        return function (y) {          return x < y ? lt : x > y ? gt : eq;        };      };    };  };};
function numShl(n1) {  return function(n2) {    return n1 << n2;  };};
function numShr(n1) {  return function(n2) {    return n1 >> n2;  };};
function numZshr(n1) {  return function(n2) {    return n1 >>> n2;  };};
function numAnd(n1) {  return function(n2) {    return n1 & n2;  };};
function numOr(n1) {  return function(n2) {    return n1 | n2;  };};
function numXor(n1) {  return function(n2) {    return n1 ^ n2;  };};
function numComplement(n) {  return ~n;};
function boolAnd(b1) {  return function(b2) {    return b1 && b2;  };};
function boolOr(b1) {  return function(b2) {    return b1 || b2;  };};
function boolNot(b) {  return !b;};
function concatString(s1) {  return function(s2) {    return s1 + s2;  };};
var $bar$bar = function (dict) {
    return dict["||"];
};
var $bar = function (dict) {
    return dict["|"];
};
var $up = function (dict) {
    return dict["^"];
};
var $greater$greater$eq = function (dict) {
    return dict[">>="];
};
var $eq$eq = function (dict) {
    return dict["=="];
};
var $less$bar$greater = function (dict) {
    return dict["<|>"];
};
var $less$greater = function (dict) {
    return dict["<>"];
};
var $less$less$less = function (dict) {
    return dict["<<<"];
};
var $greater$greater$greater = function (__dict_Semigroupoid_0) {
    return function (f) {
        return function (g) {
            return $less$less$less(__dict_Semigroupoid_0)(g)(f);
        };
    };
};
var $less$times$greater = function (dict) {
    return dict["<*>"];
};
var $less$dollar$greater = function (dict) {
    return dict["<$>"];
};
var $colon = cons;
var $div$eq = function (dict) {
    return dict["/="];
};
var $div = function (dict) {
    return dict["/"];
};
var $minus = function (dict) {
    return dict["-"];
};
var $plus$plus = function (__dict_Semigroup_1) {
    return $less$greater(__dict_Semigroup_1);
};
var $plus = function (dict) {
    return dict["+"];
};
var $times = function (dict) {
    return dict["*"];
};
var $amp$amp = function (dict) {
    return dict["&&"];
};
var $amp = function (dict) {
    return dict["&"];
};
var $percent = function (dict) {
    return dict["%"];
};
var $dollar = function (f) {
    return function (x) {
        return f(x);
    };
};
var $hash = function (x) {
    return function (f) {
        return f(x);
    };
};
var zshr = function (dict) {
    return dict.zshr;
};
var unsafeCompare = unsafeCompareImpl(LT.value)(EQ.value)(GT.value);
var unit = {};
var shr = function (dict) {
    return dict.shr;
};
var showUnit = function (_) {
    return new Show(function (_7) {
        return (function (_29) {
            return "Unit {}";
            throw new Error("Failed pattern match");
        })(_7);
    });
};
var showString = function (_) {
    return new Show(showStringImpl);
};
var showOrdering = function (_) {
    return new Show(function (_15) {
        return (function (_30) {
            if (_30 instanceof LT) {
                return "LT";
            };
            if (_30 instanceof GT) {
                return "GT";
            };
            if (_30 instanceof EQ) {
                return "EQ";
            };
            throw new Error("Failed pattern match");
        })(_15);
    });
};
var showNumber = function (_) {
    return new Show(showNumberImpl);
};
var showBoolean = function (_) {
    return new Show(function (_8) {
        return (function (_31) {
            if (_31) {
                return "true";
            };
            if (!_31) {
                return "false";
            };
            throw new Error("Failed pattern match");
        })(_8);
    });
};
var show = function (dict) {
    return dict.show;
};
var showArray = function (__dict_Show_2) {
    return new Show(showArrayImpl(show(__dict_Show_2)));
};
var shl = function (dict) {
    return dict.shl;
};
var semigroupoidArr = function (_) {
    return new Semigroupoid(function (f) {
        return function (g) {
            return function (x) {
                return f(g(x));
            };
        };
    });
};
var semigroupUnit = function (_) {
    return new Semigroup(function (_22) {
        return function (_23) {
            return (function (_32, _33) {
                return {};
                throw new Error("Failed pattern match");
            })(_22, _23);
        };
    });
};
var semigroupString = function (_) {
    return new Semigroup(concatString);
};
var semigroupArr = function (__dict_Semigroup_3) {
    return new Semigroup(function (f) {
        return function (g) {
            return function (x) {
                return $less$greater(__dict_Semigroup_3)(f(x))(g(x));
            };
        };
    });
};
var pure = function (dict) {
    return dict.pure;
};
var $$return = function (__dict_Monad_4) {
    return pure(__dict_Monad_4["__superclass_Prelude.Applicative_0"]({}));
};
var numNumber = function (_) {
    return new Num(numMod, numMul, numAdd, numSub, numDiv, numNegate);
};
var not = function (dict) {
    return dict.not;
};
var negate = function (dict) {
    return dict.negate;
};
var liftM1 = function (__dict_Monad_5) {
    return function (f) {
        return function (a) {
            return $greater$greater$eq(__dict_Monad_5["__superclass_Prelude.Bind_1"]({}))(a)(function (_0) {
                return (function (_34) {
                    var a$prime = _34;
                    return $$return(__dict_Monad_5)(f(a$prime));
                    throw new Error("Failed pattern match");
                })(_0);
            });
        };
    };
};
var liftA1 = function (__dict_Applicative_6) {
    return function (f) {
        return function (a) {
            return $less$times$greater(__dict_Applicative_6["__superclass_Prelude.Apply_0"]({}))(pure(__dict_Applicative_6)(f))(a);
        };
    };
};
var id = function (dict) {
    return dict.id;
};
var functorArr = function (_) {
    return new Functor($less$less$less(semigroupoidArr({})));
};
var flip = function (f) {
    return function (b) {
        return function (a) {
            return f(a)(b);
        };
    };
};
var eqUnit = function (_) {
    return new Eq(function (_11) {
        return function (_12) {
            return (function (_35, _36) {
                return false;
                throw new Error("Failed pattern match");
            })(_11, _12);
        };
    }, function (_9) {
        return function (_10) {
            return (function (_37, _38) {
                return true;
                throw new Error("Failed pattern match");
            })(_9, _10);
        };
    });
};
var ordUnit = function (_) {
    return new Ord(function (_) {
        return eqUnit({});
    }, function (_16) {
        return function (_17) {
            return (function (_39, _40) {
                return EQ.value;
                throw new Error("Failed pattern match");
            })(_16, _17);
        };
    });
};
var eqString = function (_) {
    return new Eq(refIneq, refEq);
};
var ordString = function (_) {
    return new Ord(function (_) {
        return eqString({});
    }, unsafeCompare);
};
var eqNumber = function (_) {
    return new Eq(refIneq, refEq);
};
var ordNumber = function (_) {
    return new Ord(function (_) {
        return eqNumber({});
    }, unsafeCompare);
};
var eqBoolean = function (_) {
    return new Eq(refIneq, refEq);
};
var ordBoolean = function (_) {
    return new Ord(function (_) {
        return eqBoolean({});
    }, function (_18) {
        return function (_19) {
            return (function (_41, _42) {
                if (!_41) {
                    if (!_42) {
                        return EQ.value;
                    };
                };
                if (!_41) {
                    if (_42) {
                        return LT.value;
                    };
                };
                if (_41) {
                    if (_42) {
                        return EQ.value;
                    };
                };
                if (_41) {
                    if (!_42) {
                        return GT.value;
                    };
                };
                throw new Error("Failed pattern match");
            })(_18, _19);
        };
    });
};
var empty = function (dict) {
    return dict.empty;
};
var $$const = function (_3) {
    return function (_4) {
        return (function (_43, _44) {
            var a = _43;
            return a;
            throw new Error("Failed pattern match");
        })(_3, _4);
    };
};
var $$void = function (__dict_Functor_8) {
    return function (fa) {
        return $less$dollar$greater(__dict_Functor_8)($$const(unit))(fa);
    };
};
var complement = function (dict) {
    return dict.complement;
};
var compare = function (dict) {
    return dict.compare;
};
var $less = function (__dict_Ord_10) {
    return function (a1) {
        return function (a2) {
            return (function (_45) {
                if (_45 instanceof LT) {
                    return true;
                };
                return false;
                throw new Error("Failed pattern match");
            })(compare(__dict_Ord_10)(a1)(a2));
        };
    };
};
var $less$eq = function (__dict_Ord_11) {
    return function (a1) {
        return function (a2) {
            return (function (_46) {
                if (_46 instanceof GT) {
                    return false;
                };
                return true;
                throw new Error("Failed pattern match");
            })(compare(__dict_Ord_11)(a1)(a2));
        };
    };
};
var $greater = function (__dict_Ord_12) {
    return function (a1) {
        return function (a2) {
            return (function (_47) {
                if (_47 instanceof GT) {
                    return true;
                };
                return false;
                throw new Error("Failed pattern match");
            })(compare(__dict_Ord_12)(a1)(a2));
        };
    };
};
var $greater$eq = function (__dict_Ord_13) {
    return function (a1) {
        return function (a2) {
            return (function (_48) {
                if (_48 instanceof LT) {
                    return false;
                };
                return true;
                throw new Error("Failed pattern match");
            })(compare(__dict_Ord_13)(a1)(a2));
        };
    };
};
var categoryArr = function (_) {
    return new Category(function (_) {
        return semigroupoidArr({});
    }, function (x) {
        return x;
    });
};
var boolLikeBoolean = function (_) {
    return new BoolLike(boolAnd, boolNot, boolOr);
};
var eqArray = function (__dict_Eq_7) {
    return new Eq(function (xs) {
        return function (ys) {
            return not(boolLikeBoolean({}))($eq$eq(eqArray(__dict_Eq_7))(xs)(ys));
        };
    }, function (xs) {
        return function (ys) {
            return eqArrayImpl($eq$eq(__dict_Eq_7))(xs)(ys);
        };
    });
};
var ordArray = function (__dict_Ord_9) {
    return new Ord(function (_) {
        return eqArray(__dict_Ord_9["__superclass_Prelude.Eq_0"]({}));
    }, function (_20) {
        return function (_21) {
            return (function (_49, _50) {
                if (_49.length === 0) {
                    if (_50.length === 0) {
                        return EQ.value;
                    };
                };
                if (_49.length === 0) {
                    return LT.value;
                };
                if (_50.length === 0) {
                    return GT.value;
                };
                if (_49.length > 0) {
                    var _54 = _49[0];
                    var _55 = _49.slice(1);
                    var xs = _55;
                    var x = _54;
                    if (_50.length > 0) {
                        var _52 = _50[0];
                        var _53 = _50.slice(1);
                        var ys = _53;
                        var y = _52;
                        return (function (_51) {
                            if (_51 instanceof EQ) {
                                return compare(ordArray(__dict_Ord_9))(xs)(ys);
                            };
                            var other = _51;
                            return other;
                            throw new Error("Failed pattern match");
                        })(compare(__dict_Ord_9)(x)(y));
                    };
                };
                throw new Error("Failed pattern match");
            })(_20, _21);
        };
    });
};
var eqOrdering = function (_) {
    return new Eq(function (x) {
        return function (y) {
            return not(boolLikeBoolean({}))($eq$eq(eqOrdering({}))(x)(y));
        };
    }, function (_13) {
        return function (_14) {
            return (function (_56, _57) {
                if (_56 instanceof LT) {
                    if (_57 instanceof LT) {
                        return true;
                    };
                };
                if (_56 instanceof GT) {
                    if (_57 instanceof GT) {
                        return true;
                    };
                };
                if (_56 instanceof EQ) {
                    if (_57 instanceof EQ) {
                        return true;
                    };
                };
                var _ = _56;
                return false;
                throw new Error("Failed pattern match");
            })(_13, _14);
        };
    });
};
var bitsNumber = function (_) {
    return new Bits(numAnd, numXor, numComplement, numShl, numShr, numZshr, numOr);
};
var asTypeOf = function (_5) {
    return function (_6) {
        return (function (_58, _59) {
            var x = _58;
            return x;
            throw new Error("Failed pattern match");
        })(_5, _6);
    };
};
var applyArr = function (_) {
    return new Apply(function (f) {
        return function (g) {
            return function (x) {
                return f(x)(g(x));
            };
        };
    }, function (_) {
        return functorArr({});
    });
};
var bindArr = function (_) {
    return new Bind(function (m) {
        return function (f) {
            return function (x) {
                return f(m(x))(x);
            };
        };
    }, function (_) {
        return applyArr({});
    });
};
var applicativeArr = function (_) {
    return new Applicative(function (_) {
        return applyArr({});
    }, $$const);
};
var monadArr = function (_) {
    return new Monad(function (_) {
        return applicativeArr({});
    }, function (_) {
        return bindArr({});
    });
};
var ap = function (__dict_Monad_14) {
    return function (f) {
        return function (a) {
            return $greater$greater$eq(__dict_Monad_14["__superclass_Prelude.Bind_1"]({}))(f)(function (_2) {
                return (function (_60) {
                    var f$prime = _60;
                    return $greater$greater$eq(__dict_Monad_14["__superclass_Prelude.Bind_1"]({}))(a)(function (_1) {
                        return (function (_61) {
                            var a$prime = _61;
                            return $$return(__dict_Monad_14)(f$prime(a$prime));
                            throw new Error("Failed pattern match");
                        })(_1);
                    });
                    throw new Error("Failed pattern match");
                })(_2);
            });
        };
    };
};
module.exports = {
    Unit: Unit, 
    LT: LT, 
    GT: GT, 
    EQ: EQ, 
    Semigroup: Semigroup, 
    BoolLike: BoolLike, 
    Bits: Bits, 
    Ord: Ord, 
    Eq: Eq, 
    Num: Num, 
    Monad: Monad, 
    Bind: Bind, 
    Alternative: Alternative, 
    Applicative: Applicative, 
    Apply: Apply, 
    Functor: Functor, 
    Show: Show, 
    Category: Category, 
    Semigroupoid: Semigroupoid, 
    unit: unit, 
    "++": $plus$plus, 
    "<>": $less$greater, 
    not: not, 
    "||": $bar$bar, 
    "&&": $amp$amp, 
    complement: complement, 
    zshr: zshr, 
    shr: shr, 
    shl: shl, 
    "^": $up, 
    "|": $bar, 
    "&": $amp, 
    ">=": $greater$eq, 
    "<=": $less$eq, 
    ">": $greater, 
    "<": $less, 
    compare: compare, 
    refIneq: refIneq, 
    refEq: refEq, 
    "/=": $div$eq, 
    "==": $eq$eq, 
    negate: negate, 
    "%": $percent, 
    "/": $div, 
    "*": $times, 
    "-": $minus, 
    "+": $plus, 
    ap: ap, 
    liftM1: liftM1, 
    "return": $$return, 
    ">>=": $greater$greater$eq, 
    "<|>": $less$bar$greater, 
    empty: empty, 
    liftA1: liftA1, 
    pure: pure, 
    "<*>": $less$times$greater, 
    "void": $$void, 
    "<$>": $less$dollar$greater, 
    show: show, 
    cons: cons, 
    ":": $colon, 
    "#": $hash, 
    "$": $dollar, 
    id: id, 
    ">>>": $greater$greater$greater, 
    "<<<": $less$less$less, 
    asTypeOf: asTypeOf, 
    "const": $$const, 
    flip: flip, 
    semigroupoidArr: semigroupoidArr, 
    categoryArr: categoryArr, 
    showUnit: showUnit, 
    showString: showString, 
    showBoolean: showBoolean, 
    showNumber: showNumber, 
    showArray: showArray, 
    functorArr: functorArr, 
    applyArr: applyArr, 
    applicativeArr: applicativeArr, 
    bindArr: bindArr, 
    monadArr: monadArr, 
    numNumber: numNumber, 
    eqUnit: eqUnit, 
    eqString: eqString, 
    eqNumber: eqNumber, 
    eqBoolean: eqBoolean, 
    eqArray: eqArray, 
    eqOrdering: eqOrdering, 
    showOrdering: showOrdering, 
    ordUnit: ordUnit, 
    ordBoolean: ordBoolean, 
    ordNumber: ordNumber, 
    ordString: ordString, 
    ordArray: ordArray, 
    bitsNumber: bitsNumber, 
    boolLikeBoolean: boolLikeBoolean, 
    semigroupUnit: semigroupUnit, 
    semigroupString: semigroupString, 
    semigroupArr: semigroupArr
};
