// Control.Monad.Eff/index.js
"use strict";
var Prelude = require("Prelude");
function returnE(a) {  return function() {    return a;  };};
function bindE(a) {  return function(f) {    return function() {      return f(a())();    };  };};
function runPure(f) {  return f();};
function untilE(f) {  return function() {    while (!f()) { }    return {};  };};
function whileE(f) {  return function(a) {    return function() {      while (f()) {        a();      }      return {};    };  };};
function forE(lo) {  return function(hi) {    return function(f) {      return function() {        for (var i = lo; i < hi; i++) {          f(i)();        }      };    };  };};
function foreachE(as) {  return function(f) {    return function() {      for (var i = 0; i < as.length; i++) {        f(as[i])();      }    };  };};
var applicativeEff = function (_) {
    return new Prelude.Applicative(function (_) {
        return applyEff({});
    }, returnE);
};
var applyEff = function (_) {
    return new Prelude.Apply(Prelude.ap(monadEff({})), function (_) {
        return functorEff({});
    });
};
var monadEff = function (_) {
    return new Prelude.Monad(function (_) {
        return applicativeEff({});
    }, function (_) {
        return bindEff({});
    });
};
var bindEff = function (_) {
    return new Prelude.Bind(bindE, function (_) {
        return applyEff({});
    });
};
var functorEff = function (_) {
    return new Prelude.Functor(Prelude.liftA1(applicativeEff({})));
};
module.exports = {
    foreachE: foreachE, 
    forE: forE, 
    whileE: whileE, 
    untilE: untilE, 
    runPure: runPure, 
    bindE: bindE, 
    returnE: returnE, 
    functorEff: functorEff, 
    applyEff: applyEff, 
    applicativeEff: applicativeEff, 
    bindEff: bindEff, 
    monadEff: monadEff
};
